### PR TITLE
[WIP] Implemented an `extract_image_patches` op using `lax` ops.

### PR DESCRIPTION
**Pull request by gehring**
An implementation of an op for extracting image patches. This implementation is a port of TensorFlow's implementation of `tf.image.extract_image_patches`.

### CODE REVIEW

**Comment by gehring** [comment]
@hawkinsp Here is a prototype of tensorflow's `extract_image_patches`. This should be an exact port of tensorflow's implementation but a bit more pythonic. I still have to go around and polish it but let me know what you think of this direction. Also, adding some tests would probably a good idea.

File: jax/ops/image/util.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+
+def get_spatial_dim(num_dims, data_format, i):
```
Comment (hawkinsp) [inline jax/ops/image/util.py]: I think the (not publicly documented) `lax.conv_dimension_numbers` has most of what you need here. Perhaps we could refactor it a bit and share that code?

Either way I wouldn't want to add a new public API for dimension numbers as part of this change.

File: jax/ops/image/extract.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+from jax import lax
+from jax.ops.image import util
+from jax.numpy import lax_numpy as np
+
+
+def extract_image_patches(image, ksizes, strides, rates, padding):
+    # replicating tensorflow's implementation
+    num_dims = image.ndim
+    num_spatial_dims = num_dims - 2
+    data_format = "NHWC"
```
Comment (hawkinsp) [inline jax/ops/image/extract.py]: You should probably let the user specify the format.

File: jax/ops/image/extract.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+from __future__ import print_function
+
+import numpy as onp
+
+from jax import lax
+from jax.ops.image import util
+from jax.numpy import lax_numpy as np
+
+
+def extract_image_patches(image, ksizes, strides, rates, padding):
```
Comment (hawkinsp) [inline jax/ops/image/extract.py]: Given that `extract_image_patches` is very similar to convolution, I would try to harmonize the API with `conv_general_dilated` as much as possible (e.g., perhaps `rhs_dilation` instead of `rates` etc, similar parameter orders, etc.)

File: jax/ops/image/extract.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    kernel_shape = [ksizes[util.get_spatial_dim(num_dims, data_format, i)]
+                    for i in range(num_spatial_dims)]
+    kernel_size = onp.product(kernel_shape)
+    kernel_shape.append(1)
+    kernel_shape.append(kernel_size * depth)
+
+    iota_kernel_shape = (kernel_size, 1, kernel_size)
+
+    conv_filter = lax.eq(
+        lax.broadcasted_iota(np.int64, iota_kernel_shape, 0),
```
Comment (hawkinsp) [inline jax/ops/image/extract.py]: I'd probably use `np.int32` here.

File: jax/ops/image/extract.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    conv_dims = list(conv.shape)
+
+    conv_dims[-1] = depth
+    conv_dims.append(kernel_size)
+    conv = lax.reshape(conv, conv_dims)
+
+    permutation = list(range(len(conv_dims)))
+    permutation[-2] = permutation[-1]
+    permutation[-1] = permutation[-2] - 1
+    # TODO(gehring): check if there is a way to transpose and reshape in one
```
Comment (hawkinsp) [inline jax/ops/image/extract.py]: You can: `reshape` can do this. But you shouldn't worry about that too much. You'd do better just to wrap a `jit` around the implementation so XLA can see the whole thing.

File: jax/ops/image/extract.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+    depth = image.shape[feature_dim]
+
+    kernel_shape = [ksizes[util.get_spatial_dim(num_dims, data_format, i)]
+                    for i in range(num_spatial_dims)]
+    kernel_size = onp.product(kernel_shape)
+    kernel_shape.append(1)
+    kernel_shape.append(kernel_size * depth)
+
+    iota_kernel_shape = (kernel_size, 1, kernel_size)
```
Comment (hawkinsp) [inline jax/ops/image/extract.py]: Shouldn't this be `(kernel_size, depth, kernel_size)`?

**Review by hawkinsp — CHANGES_REQUESTED** [review]
This looks like a great start!

File: jax/ops/image/extract.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
+# Copyright 2019 Google LLC
```
[the remark this answers is not in the document]
Comment (gehring) [reply jax/ops/image/extract.py]: Done

File: jax/ops/image/extract.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+import numpy as onp
+
+from jax import lax
+from jax.ops.image import util
+from jax.numpy import lax_numpy as np
+
+
+def extract_image_patches(image, ksizes, strides, rates, padding):
+    # replicating tensorflow's implementation
```
[the remark this answers is not in the document]
Comment (gehring) [reply jax/ops/image/extract.py]: I stopped short of finishing this because I wanted to see if we wanted to change the API first. It doesn't really make sense to me to require a specifying batch and feature dims for patch sizes, dilation and strides if we don't support !=1 (according to the original TF implementation). Should we change this to make calls simpler?

File: jax/ops/image/util.py (hunk 1 above)
Comment (gehring) [reply jax/ops/image/util.py]: Removed the util module and changed to use lax's methods.

File: jax/ops/image/extract.py (hunk 2 above)
Comment (gehring) [reply jax/ops/image/extract.py]: The original implementation had it hard coded but I don't see any reason not to allow this to be specified. I've updated the implementation.

File: jax/ops/image/extract.py (hunk 3 above)
Comment (gehring) [reply jax/ops/image/extract.py]: There isn't really a `rhs` in this case but I pattern matched as well as I could!
Let me know if you have any suggested changes. Also, the original implementation doesn't have any `lhs_dilation`. I think there is nothing inherent about this implementation that would prevent from supporting this. WDYT?

File: jax/ops/image/extract.py (hunk 4 above)
Comment (gehring) [reply jax/ops/image/extract.py]: Done

File: jax/ops/image/extract.py (hunk 6 above)
Comment (gehring) [reply jax/ops/image/extract.py]: Yes! That's what I get for only testing on images with depth 1... Now fixed.

File: jax/ops/image/extract.py (hunk 5 above)
Comment (gehring) [reply jax/ops/image/extract.py]: Sounds good! I removed the `TODO`. I'm assuming you don't expect any performance benefits from doing it with one `reshape` rather than the current form.

**Comment by lucasb-eyer** [comment]
Wow, I was just looking for this, so thanks! :) One thing to note is that this kind of operation used to be called `im2col` back in caffe/torch days, and I believe in matlab too. So it would be nice to mention this in the documentation, if just for discoverability via search.

**Comment by lucasb-eyer** [comment]
While trying this out, my repl dies with SIGABRT whenever I mistakenly pass a `np.float64` for `image`. Not sure if this should be dealt with here or it's not this function's problem. (I'm new to jax.)

**Comment by gehring** [comment]
@lucasb-eyer I wasn't able to reproduce your issue on colab (both CPU and GPU). Admittedly, I am not familiar enough with the backend to give a good guess about what might cause this error, but some additional info might help. Are you running with CUDA? Could you provide a minimal example that causes this error?

**Comment by lucasb-eyer** [comment]
I had a mistake in my error description, it crashes when I pass an integer array, not a float64.

Running with CUDA on my local workstation's GPU (a 1080Ti). The minimal example is copy-pasting the function definition from this PR into a python interpreter with jax, then running the following:

```
extract_image_patches(
    np.arange(25).reshape(1,5,5,1),
    rhs_shape=(1,2,2,1),
    window_strides=(1,1,1,1),
    padding='VALID',
    rhs_dilation=(1,1,1,1),
    data_format='NHWC')
```

This crashes, but using `np.arange(25).reshape(1,5,5,1).astype(np.float32)` works nice.

**Comment by gehring** [comment]
> I had a mistake in my error description, it crashes when I pass an integer array, not a float64.

@lucasb-eyer Someone should correct me if I am wrong but I don't believe that integer convolution is supported in JAX at the moment.

**Comment by gehring** [comment]
@hawkinsp When you have some time, could you look through my replies and changes. I'd like to hear your thoughts before making the final changes to the docstrings and function signature.

**Comment by lucasb-eyer** [comment]
Indeed, I didn't expect it to work, I just expected it not to brutally crash with a SIGABRT, though maybe my expectation is wrong? That's also why I initially wrote "Not sure if this should be dealt with here or it's not this function's problem."

**Comment by mattjj** [comment]
By the way, @hawkinsp is away right now (last week and this coming week) and asked one of us to follow up on this PR, but we haven't had the bandwidth yet. We'll get to it as soon as we can! Thanks for your work here.

**Comment by gehring** [comment]
@mattjj No problem at all! Thanks for the heads up!

**Comment by MostafaDehghani** [comment]
This is great @gehring. I wonder if you have an idea on the simplest way of having the inverse of this, i.e. from patches to the original image?

**Comment by hawkinsp** [comment]
@gehring I haven't forgotten about reviewing this, but I've been out of office a lot. I'll take a look when I can.

@MostafaDehghani Usually when you are looking for that kind of transpose operation, you can use the reverse-mode gradient, or VJP.

`jax.vjp` should let you do what you need here: use `jax.vjp` on `extract_image_patches` with a cotangent containing the image patches to sum. If there are overlapping image patches, the VJP will sum the contributions all of them. To take the mean, which I assume is what you want, you can use `jax.vjp` with a cotangent of all 1s to count the number of contributions to each pixel, then divide the sum by the count. Hope that helps!

**Comment by MostafaDehghani** [comment]
Hi, 

Just for the record, wanted to post this small example of how to reconstruct the original image given the output of extract_image_patches using  jax.vjp. 

```
im = np.arange(36).reshape(-1,6,6,1).astype(np.float32)
patch_height = patch_width = 2

def i2p(im):
  return extract_image_patches(
          lhs=im.astype(np.float64),
          rhs_shape=(1,patch_height, patch_width,1),
          window_strides=(1,patch_height, patch_width,1), 
          padding='VALID',
          rhs_dilation=(1,1,1,1),
          data_format='NHWC')

def p2i(im_patched):
    _, fn_vjp = jax.vjp(i2p, np.ones(im.shape))
    overlap_count = fn_vjp(np.ones_like(im_patched))[0]
    return fn_vjp(im_patched)[0] / overlap_cou

im_patched = i2p(im).squeeze()
recunstructed_im = p2i(im_patched)
assert np.array_equal(recunstructed_im, im)
```
Thanks  again @hawkinsp for the help! :)

**Comment by MostafaDehghani** [comment]
I know extract_image_patches in pytorch, etc. returns <bs,  lhs_shape[1],  lhs_shape[2],   rhs_shape[1] * rhs_shape[2] *C>``` , but I wish it was ```<bs,  lhs_shape[1],  lhs_shape[2],   rhs_shape[1],  rhs_shape[2],  C>``` instead and just leave it to the user if they want to do the space-to-depth part!

**Comment by gehring** [comment]
This PR is no longer relevant given a more polished version is actively being worked on (#4459).
